src/pass/autodiff.cc

```diff
@@ -0,0 +1,539 @@
+                                                      Map<Tensor, Map<Tensor, Tensor>> summands) {
+  auto n = make_node<DifferentiationResultNode>();
+  n->result = std::move(result);
+  n->adjoints = adjoints;
+  n->adjoint_summands = summands;
+  return DifferentiationResult(n);
+}
+
+TVM_STATIC_IR_FUNCTOR(IRPrinter, vtable)
+.set_dispatch<DifferentiationResultNode>([](const DifferentiationResultNode* r, IRPrinter* p) {
+    p->stream << "DifferentiationResult(result=" << r->result
+              << ", adjoints=" << r->adjoints
+              << ", adjoint_summands=" << r->adjoint_summands << ')';
+  });
+
+TVM_REGISTER_NODE_TYPE(DifferentiationResultNode);
+
+
+#define NOT_IMPLEMENTED \
+  { LOG(FATAL) << "Derivative of this expr is not implemented: " << e; throw; }
+
+/*! \brief Differentiate an expression wrt a variable or a tensor element */
+class JacobianMutator : public IRMutator {
+ public:
+  /*!
+   * \brief Differentiate wrt `input(indices)`.
+   * \param input The input tensor.
+   * \param indices The indices of the element with respect to which to differentiate.
+   */
+  explicit JacobianMutator(Tensor input, Array<Expr> indices)
+    : input_(input), indices_(indices) {}
+  /*!
+   * \brief Differentiate wrt the input variable.
+   * \param input The input variable.
+   */
+  explicit JacobianMutator(VarExpr input)
+    : input_var_(input) {}
+
+  virtual Expr Mutate(Expr e) {
+    if (e.type().is_int() || e.type().is_uint()) {
+      // Assume that the derivative of any integer expression is always 0
+      return make_zero(e.type());
+    } else {
+      return IRMutator::Mutate(e);
+    }
+  }
+
+  Expr Mutate_(const Variable* op, const Expr& e) {
+    if (input_var_.operator->() && input_var_.get() == op && op->type.is_float()) {
+      return FloatImm::make(op->type, 1.0);
+    } else {
+      return make_zero(op->type);
+    }
+  }
+
+  Expr Mutate_(const Load* op, const Expr& e) NOT_IMPLEMENTED
+  Expr Mutate_(const Let* op, const Expr& e) NOT_IMPLEMENTED
+
+  Expr Mutate_(const Call* op, const Expr& e) {
+    if (op->call_type == Call::CallType::Halide) {
+      if (input_.operator->() && op->func.same_as(input_->op) &&
+          op->value_index == input_->value_index) {
+        Expr condition = const_true();
+        for (size_t i = 0; i < input_.ndim(); ++i) {
+          condition = And::make(condition, EQ::make(indices_[i], op->args[i]));
+        }
+        return Cast::make(op->type, condition);
+      } else {
+        return make_zero(op->type);
+      }
+    } else if (op->call_type == Call::CallType::PureIntrinsic) {
+      static std::unordered_set<std::string> piecewise_const = {"floor", "ceil", "trunc", "round"};
+      if (op->name == "exp") {
+        return Mul::make(Mutate(op->args[0]), e);
+      } else if (op->name == "log") {
+        return Div::make(Mutate(op->args[0]), op->args[0]);
+      } else if (op->name == "sigmoid") {
```

Comment:
Yes, since softmax is not elementwise, we can't express it here, so I think it should be done through the gradient overriding mechanism (or on the Relay level).

Reply:
I agree.   Although softmax can be inside a large express tree (e.g. focalloss) and as such, we'd want to autodiff that express tree (using the Jacobian logic you have there), but when coming up the softmax operator, using the relay overridden softmax gradient.

Reply:
I think there are several problems:
- How to override differentiation rules for some specific tensor: in my implementation there is a mechanism for this, e.g. the option [override](https://github.com/sgrechanik-h/tvm/blob/a342e5e93fa067b135f918cc0d9f2fef67dba1bd/python/tvm/autodiff.py#L74) which can be used as described [here](https://github.com/sgrechanik-h/tvm/blob/a342e5e93fa067b135f918cc0d9f2fef67dba1bd/tutorials/language/autodiff_basics.py#L132)
- How to detect the operation for which we want to override gradients, in this case softmax. The overriding mechanism simply compares tensors by reference, but this equality seems to be too fine-grained in practice, because `sortmax(A)` and `softmax(B)` are two different tensors on the tensor expression level. I think the practical solution would be to perform gradient overriding by the tensor's tag or name, however currently it is not implemented.
- How to reuse gradients implemented in Relay. I don't know, but I think it should be relatively straightforward to lower a Relay graph representing a gradient into a tensor expression (however I have no idea what to do with schedules).

Reply:
Thanks for the reply!  Took me a while to understand but I like your clever design that when override is specified, **fdiff** is essentially a wrapper.  That is, depending on whether the **out** 
 tensor is specified in the override or not, one would either use the default fdiff, or the one specified inside the override dictionary.  However, I found your example below unintuitive: 
```
          # a generalization of my_fdiff which works for non-immediate dependencies
        # this is necessary because z1 is not an immediate dep of z2 because of padding
        def my_diff(out, inputs, head):
            return tvm.differentiate(out, inputs, head, fdiff=my_fdiff)

        # using a custom differentiation function only for z2
        res = tvm.differentiate(y, [w1, w2], override={z2: ([z1, w2], my_diff)})
```

If I write my own custom fdiff for **z2** w.r.t. z1, or w2, I'd prefer that my_diff returning a tensor (i.e. a tvm.compute that is the custom gradient)  instead of DifferentiationResult.  (not sure how to construct such object in python...)

Reply:
Actually,  I jut realize that I can indeed do so, by putting the custom gradient tensors into a list and returning that from my_diff.  DifferentiationResult is simply a list of tensors!